File: dbxml/examples/perl/gettingStarted/queryForDocumentValue.pl


use strict;
use warnings;
use Sleepycat::DbXml 'simple';


sub usage()
{
    warn <<EOM;
This program performs various XPath queries against a DBXML container and then
retrieves information of interest from the retrieved document(s).  You should
pre-load the container using loadExamplesData.[sh|cmd] before running this
example.  You are only required to pass this command the path location of the
database environment that you specified when you pre-loaded the examples data:

    -h <dbenv directory>
EOM
        exit( -1 );
}

sub getValue($$$$)
{
    my $mgr         = shift ;
    my $document    = shift ;
    my $XPath       = shift ;
    my $context     = shift ;

    #///////////////////////////////////////////////////////////////////////
    #//////////    Return specific information from a document. ////////////
    #//////////   !!!!!! Assumes a result set of size 1 !!!!!!! ////////////
    #///////////////////////////////////////////////////////////////////////


    #  Exception handling omitted....

    # Perform the query
    my $doc_expr = $mgr->prepare($XPath, $context);
    my $result = $doc_expr->execute(new XmlValue($document), $context);

    # We require a result set size of 1.
    die "Result set != 1" unless $result->size() == 1 ;

    # Get the value. If we allowed the result set to be larger than size 1,
    # we would have to loop through the results, processing each as is
    # required by our application.
    my $value;
    $result->next($value);

    return $value;
        
}

sub getDetails($$$$)
{
    my $mgr = shift ;
    my $container = shift ;
    my $query = shift ;
    my $context = shift ;

    #///////////////////////////////////////////////////////////////////////
    #/////  Performs an query (in context) against the               ///////
    #/////  provided container.                                      ///////
    #///////////////////////////////////////////////////////////////////////


    #//// some defensive code eliminated for clarity //

    #  Perform the query. Result type is by default Result Document
    my $fullQuery = "collection('" .$container->getName() . "')$query";
    eval {
        print "Exercising query '$fullQuery' \n";
        print "Return to continue: ";
        <>;

        my $results = $mgr->query($fullQuery, $context );
        my $value = new XmlValue();
        print "\n\tProduct : Price : Inventory Level\n";
        while( $results->next($value) ) {
            #  Retrieve the value as a document
            my $theDocument = $value->asDocument();

            #  Obtain information of interest from the document. Note that the 
            #  wildcard in the query expression allows us to not worry about what
            #  namespace this document uses.
            my $item = getValue($mgr, $theDocument, "/*/product/text()", $context);
            my $price = getValue($mgr, $theDocument, "/*/inventory/price/text()", $context);
            my $inventory = getValue($mgr, $theDocument, "/*/inventory/inventory/text()", $context);

            print "\t$item : $price : $inventory\n";

        }
        print "\n";
        print $results->size() . " objects returned for expression '$fullQuery'\n" ;
    };
    # Catches XmlException
    if (my $e = catch std::exception)
    {
        warn "Query $fullQuery failed\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Query $fullQuery failed\n";
        warn $@;
        exit( -1 );
    }    
}

my $path2DbEnv;
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/^-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
     if ! $path2DbEnv ;

 # Open a container in the db environment
my $env = new DbEnv(0);
$env->set_cachesize(0, 64 * 1024 * 1024, 1);
$env->open($path2DbEnv,
    Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
my $mgr = new XmlManager($env);

my $containerTxn = $mgr->createTransaction();
my $container = $mgr->openContainer($containerTxn, $theContainer);
$containerTxn->commit();

# Create a context and declare the namespaces
my $context = $mgr->createQueryContext();
$context->setNamespace('fruits'     => 'http://groceryItem.dbxml/fruits');
$context->setNamespace('vegetables' => 'http://groceryItem.dbxml/vegetables');
$context->setNamespace('desserts'   => 'http://groceryItem.dbxml/desserts');

# get details on Zulu Nuts
getDetails($mgr, $container, "/fruits:item/product[text() = 'Zulu Nut']", $context);

# get details on all fruits that start with 'A'
getDetails($mgr, $container, "/vegetables:item/product[starts-with(text(),'A')]", $context);

File: dbxml/examples/java/gettingStarted/queryForDocumentValue.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;
import com.sleepycat.dbxml.*;

class queryForDocumentValue
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String collection = "collection('namespaceExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program retrieves DBXML documents and then retrieves information of\n";
	usageMessage += "interest from the retrieved document(s). You should run exampleLoadContainer\n";
	usageMessage += "before running this example. You are only required to pass this command the path\n";
	usageMessage += "location of the database environment that you specified when you loaded the\n";
	usageMessage += "examples data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.queryForDocumentValue -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    private static String getValue( XmlManager mgr, XmlDocument document,
                                    String query, XmlQueryContext context )
	throws Throwable {
        ///////////////////////////////////////////////////////////////////////////
        ////////    Return specific information from a document. //////////////////
        ////////          Assumes a result set of size 1         //////////////////
        ///////////////////////////////////////////////////////////////////////////

        //Perform the query
	// The document provides the context for the query, so neither
	// collection() nor doc() needs to be part of the query.
        XmlQueryExpression queryExpr = mgr.prepare(query, context);
	XmlValue docVal = new XmlValue(document);
        XmlResults result = queryExpr.execute(docVal, context, null);

        //We require a result set size of 1.
        if (result.size() != 1) {
	    System.err.println("Error!  query '" + query +
			       "' returned a result size size != 1");
	    throw new Exception( "getValue found result set not equal to 1.");
        }

        //Get the value. If we allowed the result set to be larger than size 1,
        // we would have to loop through the results, processing each as is
        // required by our application.
        XmlValue value = result.next();
	String retVal = value.asString();

	result.delete();
	queryExpr.delete();

        return retVal;
    }

    private static void getDetails(XmlManager mgr,
				   String query, XmlQueryContext context)
	throws Throwable {
	String fullQuery = collection + query;
	//Perform a single query against the referenced container using
	// the referenced context.
	System.out.println("Exercising query: '" + fullQuery + "'.");
	System.out.println("Return to continue: ");
	System.in.read();

	//Perform the query
	XmlResults results = mgr.query(fullQuery, context, null);

	System.out.println("\n\tProduct : Price : Inventory Level\n");

	XmlValue value = results.next();
	while (value != null) {

            /// Retrieve the value as a document
            XmlDocument theDocument = value.asDocument();

            //Obtain information of interest from the document. Note that the
            //  wildcard in the query expression allows us to not worry about what
            //  namespace this document uses.
            String item = getValue(mgr, theDocument, "fn:string(/*/product)",
				   context);
            String price = getValue(mgr, theDocument,
				    "fn:string(/*/inventory/price)", context);
            String inventory = getValue(mgr, theDocument,
					"fn:string(/*/inventory/inventory)", context);

            System.out.println("\t" + item + " : " + price + " : " + inventory);
            value = results.next();
	}
	System.out.println(results.size() + " results returned for query '" +
			   fullQuery + "'.");
	results.delete();
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;

	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File(args[++i]);
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	//This example does not use transactions
	myDbEnv env = null;
	XmlContainer openedContainer = null;
	try {
	    //Create an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    //Open a container in that environment
	    openedContainer =
		theMgr.openContainer(theContainer);

            //Create a context and declare the namespaces
            XmlQueryContext context = theMgr.createQueryContext();
            context.setNamespace("fruits", "http://groceryItem.dbxml/fruits");
            context.setNamespace("vegetables", "http://groceryItem.dbxml/vegetables");
            context.setNamespace("desserts", "http://groceryItem.dbxml/desserts");

            //Get details on Zulu Nuts
            getDetails(theMgr,
		       "/fruits:item[fn:string(product) = 'Zulu Nut']", context);

            //Get details on all fruits that start with 'A'
            getDetails(theMgr,
		       "/vegetables:item[starts-with(fn:string(product),'A')]",
		       context);

	} catch (Exception e) {

            System.err.println("Error performing query against " + theContainer);
            System.err.println("   Message: " + e.getMessage());
            throw e;

	}
	finally {
	    cleanup(env, openedContainer);
	}
    } //End main
}
